From e40327b15f52c34709fe4d0c851174ddec1a3c51 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 5 Mar 2022 20:50:15 +0100
Subject: [PATCH 04/19] winex11.drv: Use set_focus in FocusIn events to
 activate focused window.

Setting the X11 input focus to the new focus window after calling
SetForegroundWindow, if necessary.

CW-Bug-Id: #19553
CW-Bug-Id: #19644
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=2155

 100.0% dlls/winex11.drv/
diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 45f3e54dde6..7ab3d0e82c4 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -591,7 +591,7 @@ static void set_input_focus( struct x11drv_win_data *data )
 static void set_focus( Display *display, HWND hwnd, Time time )
 {
     HWND focus;
-    Window win;
+    Window window;
     GUITHREADINFO threadinfo;
 
     old_active = NtUserGetForegroundWindow();
@@ -602,12 +602,12 @@ static void set_focus( Display *display, HWND hwnd, Time time )
     focus = threadinfo.hwndFocus;
     if (!focus) focus = threadinfo.hwndActive;
     if (focus) focus = NtUserGetAncestor( focus, GA_ROOT );
-    win = X11DRV_get_whole_window(focus);
+    if (focus == hwnd && !use_take_focus) return;
 
-    if (win)
+    if ((window = X11DRV_get_whole_window( focus )))
     {
-        TRACE( "setting focus to %p (%lx) time=%ld\n", focus, win, time );
-        XSetInputFocus( display, win, RevertToParent, time );
+        TRACE( "setting focus to %p/%lx, time %lu\n", focus, window, time );
+        XSetInputFocus( display, window, RevertToParent, time );
     }
 }
 
@@ -809,8 +809,9 @@ static BOOL X11DRV_FocusIn( HWND hwnd, XEvent *xev )
         if (!hwnd) hwnd = get_active_window();
         if (!hwnd) hwnd = x11drv_thread_data()->last_focus;
         if (hwnd && can_activate_window(hwnd)) set_focus( event->display, hwnd, CurrentTime );
+        return TRUE;
     }
-    else NtUserSetForegroundWindow( hwnd );
+    else set_focus( event->display, hwnd, CurrentTime );
     return TRUE;
 }
 
-- 
2.43.0

